In [112]:
import csv
import re
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import probscale

plt.style.use('ggplot')

%config InlineBackend.figure_format = 'retina'
%matplotlib inline

In [113]:
import seaborn as sns

In [114]:
from sklearn.linear_model import LinearRegression

# Read Data into DataFrame

In [115]:
df = pd.read_csv("Data.csv")
data = list(df)

In [116]:
# Data is now in a list of lists format
data[:2]

['Order', 'PID']

In [117]:
# Create headings list for DataFrame
headings = data[0]

In [118]:
# Need to delete all spaces in the headings
headings = [re.sub(r'\s+', '', item) for item in data]

In [119]:
pd.set_option('display.max_columns', None)  # Show all columns in DataFrame display


df = df.rename(columns=dict(zip(df.columns, headings)))

In [120]:
df.drop(['Order'], axis=1, inplace=True)

In [121]:
df

PID  MSSubClass MSZoning  LotFrontage  LotArea Street Alley  \
0     526301100          20       RL        141.0    31770   Pave   NaN   
1     526350040          20       RH         80.0    11622   Pave   NaN   
2     526351010          20       RL         81.0    14267   Pave   NaN   
3     526353030          20       RL         93.0    11160   Pave   NaN   
4     527105010          60       RL         74.0    13830   Pave   NaN   
...         ...         ...      ...          ...      ...    ...   ...   
2925  923275080          80       RL         37.0     7937   Pave   NaN   
2926  923276100          20       RL          NaN     8885   Pave   NaN   
2927  923400125          85       RL         62.0    10441   Pave   NaN   
2928  924100070          20       RL         77.0    10010   Pave   NaN   
2929  924151050          60       RL         74.0     9627   Pave   NaN   

     LotShape LandContour Utilities LotConfig LandSlope Neighborhood  \
0         IR1         Lvl    AllPub    Corner       Gtl        NAmes   
1         Reg         Lvl    AllPub    Inside       Gtl        NAmes   
2         IR1         Lvl    AllPub    Corner       Gtl        NAmes   
3         Reg         Lvl    AllPub    Corner       Gtl        NAmes   
4         IR1         Lvl    AllPub    Inside       Gtl      Gilbert   
...       ...         ...       ...       ...       ...          ...   
2925      IR1         Lvl    AllPub   CulDSac       Gtl      Mitchel   
2926      IR1         Low    AllPub    Inside       Mod      Mitchel   
2927      Reg         Lvl    AllPub    Inside       Gtl      Mitchel   
2928      Reg         Lvl    AllPub    Inside       Mod      Mitchel   
2929      Reg         Lvl    AllPub    Inside       Mod      Mitchel   

     Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  \
0          Norm       Norm     1Fam     1Story            6            5   
1         Feedr       Norm     1Fam     1Story            5            6   
2          Norm       Norm     1Fam     1Story            6            6   
3          Norm       Norm     1Fam     1Story            7            5   
4          Norm       Norm     1Fam     2Story            5            5   
...         ...        ...      ...        ...          ...          ...   
2925       Norm       Norm     1Fam       SLvl            6            6   
2926       Norm       Norm     1Fam     1Story            5            5   
2927       Norm       Norm     1Fam     SFoyer            5            5   
2928       Norm       Norm     1Fam     1Story            5            5   
2929       Norm       Norm     1Fam     2Story            7            5   

      YearBuilt  YearRemod/Add RoofStyle RoofMatl Exterior1st Exterior2nd  \
0          1960           1960       Hip  CompShg     BrkFace     Plywood   
1          1961           1961     Gable  CompShg     VinylSd     VinylSd   
2          1958           1958       Hip  CompShg     Wd Sdng     Wd Sdng   
3          1968           1968       Hip  CompShg     BrkFace     BrkFace   
4          1997           1998     Gable  CompShg     VinylSd     VinylSd   
...         ...            ...       ...      ...         ...         ...   
2925       1984           1984     Gable  CompShg     HdBoard     HdBoard   
2926       1983           1983     Gable  CompShg     HdBoard     HdBoard   
2927       1992           1992     Gable  CompShg     HdBoard     Wd Shng   
2928       1974           1975     Gable  CompShg     HdBoard     HdBoard   
2929       1993           1994     Gable  CompShg     HdBoard     HdBoard   

     MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0         Stone       112.0        TA        TA     CBlock       TA       Gd   
1           NaN         0.0        TA        TA     CBlock       TA       TA   
2       BrkFace       108.0        TA        TA     CBlock       TA       TA   
3           NaN         0.0        Gd        TA     CBlock       TA       TA   
4       

In [122]:
# DataFRame is full of 'NA' pr blank entries that need to be turned to proper NAN
df = df.replace(['NA', ''], np.NaN)

In [123]:
df

PID  MSSubClass MSZoning  LotFrontage  LotArea Street Alley  \
0     526301100          20       RL        141.0    31770   Pave   NaN   
1     526350040          20       RH         80.0    11622   Pave   NaN   
2     526351010          20       RL         81.0    14267   Pave   NaN   
3     526353030          20       RL         93.0    11160   Pave   NaN   
4     527105010          60       RL         74.0    13830   Pave   NaN   
...         ...         ...      ...          ...      ...    ...   ...   
2925  923275080          80       RL         37.0     7937   Pave   NaN   
2926  923276100          20       RL          NaN     8885   Pave   NaN   
2927  923400125          85       RL         62.0    10441   Pave   NaN   
2928  924100070          20       RL         77.0    10010   Pave   NaN   
2929  924151050          60       RL         74.0     9627   Pave   NaN   

     LotShape LandContour Utilities LotConfig LandSlope Neighborhood  \
0         IR1         Lvl    AllPub    Corner       Gtl        NAmes   
1         Reg         Lvl    AllPub    Inside       Gtl        NAmes   
2         IR1         Lvl    AllPub    Corner       Gtl        NAmes   
3         Reg         Lvl    AllPub    Corner       Gtl        NAmes   
4         IR1         Lvl    AllPub    Inside       Gtl      Gilbert   
...       ...         ...       ...       ...       ...          ...   
2925      IR1         Lvl    AllPub   CulDSac       Gtl      Mitchel   
2926      IR1         Low    AllPub    Inside       Mod      Mitchel   
2927      Reg         Lvl    AllPub    Inside       Gtl      Mitchel   
2928      Reg         Lvl    AllPub    Inside       Mod      Mitchel   
2929      Reg         Lvl    AllPub    Inside       Mod      Mitchel   

     Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  \
0          Norm       Norm     1Fam     1Story            6            5   
1         Feedr       Norm     1Fam     1Story            5            6   
2          Norm       Norm     1Fam     1Story            6            6   
3          Norm       Norm     1Fam     1Story            7            5   
4          Norm       Norm     1Fam     2Story            5            5   
...         ...        ...      ...        ...          ...          ...   
2925       Norm       Norm     1Fam       SLvl            6            6   
2926       Norm       Norm     1Fam     1Story            5            5   
2927       Norm       Norm     1Fam     SFoyer            5            5   
2928       Norm       Norm     1Fam     1Story            5            5   
2929       Norm       Norm     1Fam     2Story            7            5   

      YearBuilt  YearRemod/Add RoofStyle RoofMatl Exterior1st Exterior2nd  \
0          1960           1960       Hip  CompShg     BrkFace     Plywood   
1          1961           1961     Gable  CompShg     VinylSd     VinylSd   
2          1958           1958       Hip  CompShg     Wd Sdng     Wd Sdng   
3          1968           1968       Hip  CompShg     BrkFace     BrkFace   
4          1997           1998     Gable  CompShg     VinylSd     VinylSd   
...         ...            ...       ...      ...         ...         ...   
2925       1984           1984     Gable  CompShg     HdBoard     HdBoard   
2926       1983           1983     Gable  CompShg     HdBoard     HdBoard   
2927       1992           1992     Gable  CompShg     HdBoard     Wd Shng   
2928       1974           1975     Gable  CompShg     HdBoard     HdBoard   
2929       1993           1994     Gable  CompShg     HdBoard     HdBoard   

     MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0         Stone       112.0        TA        TA     CBlock       TA       Gd   
1           NaN         0.0        TA        TA     CBlock       TA       TA   
2       BrkFace       108.0        TA        TA     CBlock       TA       TA   
3           NaN         0.0        Gd        TA     CBlock       TA       TA   
4       

In [124]:
# Identify the numeric data columns from the data description document
cols = ['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemod/Add', 
        'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 
        'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 
        'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 
        'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold', 
        'SalePrice']

In [125]:
df[cols] = df[cols].astype('float')

In [126]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PID            2930 non-null   int64  
 1   MSSubClass     2930 non-null   float64
 2   MSZoning       2930 non-null   object 
 3   LotFrontage    2440 non-null   float64
 4   LotArea        2930 non-null   float64
 5   Street         2930 non-null   object 
 6   Alley          198 non-null    object 
 7   LotShape       2930 non-null   object 
 8   LandContour    2930 non-null   object 
 9   Utilities      2930 non-null   object 
 10  LotConfig      2930 non-null   object 
 11  LandSlope      2930 non-null   object 
 12  Neighborhood   2930 non-null   object 
 13  Condition1     2930 non-null   object 
 14  Condition2     2930 non-null   object 
 15  BldgType       2930 non-null   object 
 16  HouseStyle     2930 non-null   object 
 17  OverallQual    2930 non-null   float64
 18  OverallC

In [127]:
df.describe()

PID   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  2.930000e+03  2930.000000  2440.000000    2930.000000  2930.000000   
mean   7.144645e+08    57.387372    69.224590   10147.921843     6.094881   
std    1.887308e+08    42.638025    23.365335    7880.017759     1.411026   
min    5.263011e+08    20.000000    21.000000    1300.000000     1.000000   
25%    5.284770e+08    20.000000    58.000000    7440.250000     5.000000   
50%    5.354536e+08    50.000000    68.000000    9436.500000     6.000000   
75%    9.071811e+08    70.000000    80.000000   11555.250000     7.000000   
max    1.007100e+09   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemod/Add   MasVnrArea   BsmtFinSF1  \
count  2930.000000  2930.000000    2930.000000  2907.000000  2929.000000   
mean      5.563140  1971.356314    1984.266553   101.896801   442.629566   
std       1.111537    30.245361      20.860286   179.112611   455.590839   
min       1.000000  1872.000000    1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000    1965.000000     0.000000     0.000000   
50%       5.000000  1973.000000    1993.000000     0.000000   370.000000   
75%       6.000000  2001.000000    2004.000000   164.000000   734.000000   
max       9.000000  2010.000000    2010.000000  1600.000000  5644.000000   

        BsmtFinSF2    BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  \
count  2929.000000  2929.000000  2929.000000  2930.000000  2930.000000   
mean     49.722431   559.262547  1051.614544  1159.557679   335.455973   
std     169.168476   439.494153   440.615067   391.890885   428.395715   
min       0.000000     0.000000     0.000000   334.000000     0.000000   
25%       0.000000   219.000000   793.000000   876.250000     0.000000   
50%       0.000000   466.000000   990.000000  1084.000000     0.000000   
75%       0.000000   802.000000  1302.000000  1384.000000   703.750000   
max    1526.000000  2336.000000  6110.000000  5095.000000  2065.000000   

       LowQualFinSF    GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath  \
count   2930.000000  2930.000000   2928.000000   2928.000000  2930.000000   
mean       4.676792  1499.690444      0.431352      0.061134     1.566553   
std       46.310510   505.508887      0.524820      0.245254     0.552941   
min        0.000000   334.000000      0.000000      0.000000     0.000000   
25%        0.000000  1126.000000      0.000000      0.000000     1.000000   
50%        0.000000  1442.000000      0.000000      0.000000     2.000000   
75%        0.000000  1742.750000      1.000000      0.000000     2.000000   
max     1064.000000  5642.000000      3.000000      2.000000     4.000000   

          HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  \
count  2930.000000   2930.000000   2930.000000   2930.000000  2930.000000   
mean      0.379522      2.854266      1.044369      6.443003     0.599317   
std       0.502629      0.827731      0.214076      1.572964     0.647921   
min       0.000000      0.000000      0.000000      2.000000     0.000000   
25%       0.000000      2.000000      1.000000      5.000000     0.000000   
50%       0.000000      3.000000      1.000000      6.000000     1.000000   
75%       1.000000      3.000000      1.000000      7.000000     1.000000   
max       2.000000      8.000000      3.000000     15.000000     4.000000   

       GarageYrBlt   GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  \
count  2771.000000  2929.000000  2929.000000  2930.000000  2930.000000   
mean   1978.132443     1.766815   472.819734    93.751877    47.533447   
std      25.528411     0.760566   215.046549   126.361562    67.483400   
min    1895.000000     0.000000     0.000000     0.000000     0.000000   
25%    1960.000000     1.000000   320.000000     0.000000     0.000000   
50%    1979.000000     2.000000   480.000000     0.000000    27.000000   
75%    2002.000000     2.000000   576.000000   168.000000    70.

# Data Engineering

In [128]:
# Dropping those variables with less than 300 non-null values
df.drop(['Alley', 'PoolQC', 'MiscFeature'], axis=1, inplace=True)
df.shape

(2930, 78)

In [129]:
df.MSZoning.unique() # C denotes commercial properties

array(['RL', 'RH', 'FV', 'RM', 'C (all)', 'I (all)', 'A (agr)'],
      dtype=object)

In [130]:
# Delete all data with MSZoning = commercial, agriculture and industrial as these are not residential units
df = df[(df.MSZoning != 'C (all)') & (df.MSZoning != 'I (all)') & (df.MSZoning != 'A (agr)')]

In [131]:
# Notice a large number of null values in 'LotFrontage'. Examine the 'Lots' grouping.
df_lots = df[['LotFrontage', 'LotArea', 'LotConfig', 'LotShape']]
grouped_lots = df_lots.groupby(['LotShape'])
mean_without_lotconfig = grouped_lots[['LotFrontage', 'LotArea']].mean()


mean_without_lotconfig 

LotFrontage       LotArea
LotShape                           
IR1         74.768760  11506.414959
IR2         67.437500  18913.065789
IR3        117.636364  32064.187500
Reg         66.821387   8849.153301

In [132]:
grouped_lots2 = df_lots.groupby(['LotConfig'])
grouped_lots2.count()

LotFrontage  LotArea  LotShape
LotConfig                                
Corner             399      503       503
CulDSac             92      180       180
FR2                 65       85        85
FR3                 10       14        14
Inside            1848     2119      2119

In [133]:
# Generate 'Lots' group where there are null 'LotFrontage' values
df_LotFrontage_NA = df_lots.loc[(df['LotFrontage'].isnull())]
df_LotFrontage_NA.columns

Index(['LotFrontage', 'LotArea', 'LotConfig', 'LotShape'], dtype='object')

In [134]:
df_LotFrontage_NA.LotFrontage.isnull().sum()

487

In [135]:
# A reasonable assumption is that LotFrontage is linked to LotConfig and LotShape, and
# the other 'Lot' variables have all observations
# So I will replace all NA in 'LotFrontage' with its mean based on the corresponding 'LotShape', which is indexed at 3
df_LotFrontage_NA['LotFrontage'] = df_LotFrontage_NA.apply(lambda x: 74.7688 if (x[3] == 'IR1') 
                                                           else x[0], axis=1)
df_LotFrontage_NA['LotFrontage'] = df_LotFrontage_NA.apply(lambda x: 67.4375 if (x[3] == 'IR2') 
                                                           else x[0], axis=1)
df_LotFrontage_NA['LotFrontage'] = df_LotFrontage_NA.apply(lambda x: 117.6364 if (x[3] == 'IR3') 
                                                           else x[0], axis=1)
df_LotFrontage_NA['LotFrontage'] = df_LotFrontage_NA.apply(lambda x: 66.8214 if (x[3] == 'Reg') 
                                                           else x[0], axis=1)

C:\Users\Ahsan Salman Yousaf\AppData\Local\Temp\ipykernel_21368\3852392954.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_LotFrontage_NA['LotFrontage'] = df_LotFrontage_NA.apply(lambda x: 74.7688 if (x[3] == 'IR1')
C:\Users\Ahsan Salman Yousaf\AppData\Local\Temp\ipykernel_21368\3852392954.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_LotFrontage_NA['LotFrontage'] = df_LotFrontage_NA.apply(lambda x: 67.4375 if (x[3] == 'IR2')
C:\Users\Ahsan Salman Yousaf\AppData\Local\Temp\ipykernel_21368\38

In [136]:
LotFront_fills = df_LotFrontage_NA.LotFrontage
LotFront_fills.count()

487

In [137]:
df.LotFrontage.isnull().sum()

487

In [138]:
# Filling the 'LotFrontage' null values with the LotFront_fills series in the given order of the data
df.loc[df.LotFrontage.isnull(), 'LotFrontage'] = LotFront_fills
df.LotFrontage.isnull().sum()

0

In [139]:
# Adding structure age variable depending if there was a major remodeling
df['Age'] = df.apply(lambda x: x['YrSold']-x['YearBuilt'] if (x['YearBuilt']<x['YearRemod/Add']) 
                                                           else (x['YrSold']-x['YearRemod/Add']), axis=1)

C:\Users\Ahsan Salman Yousaf\AppData\Local\Temp\ipykernel_21368\3016910190.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Age'] = df.apply(lambda x: x['YrSold']-x['YearBuilt'] if (x['YearBuilt']<x['YearRemod/Add'])


### Creating variables on property area/size

In [140]:
# One of the chief factors affecting house prices in the size of the property.
# Here we examine the basement square footage variables. 
df[['BsmtFinSF1', 'BsmtFinSF2', 'TotalBsmtSF', 'BsmtUnfSF']].head()

BsmtFinSF1  BsmtFinSF2  TotalBsmtSF  BsmtUnfSF
0       639.0         0.0       1080.0      441.0
1       468.0       144.0        882.0      270.0
2       923.0         0.0       1329.0      406.0
3      1065.0         0.0       2110.0     1045.0
4       791.0         0.0        928.0      137.0

In [141]:
# Here we examine the ground square footage variables. The first column is obviously the sum of the other two.
df[['GrLivArea', '1stFlrSF', '2ndFlrSF']].head()

GrLivArea  1stFlrSF  2ndFlrSF
0     1656.0    1656.0       0.0
1      896.0     896.0       0.0
2     1329.0    1329.0       0.0
3     2110.0    2110.0       0.0
4     1629.0     928.0     701.0

In [142]:
df.GrLivArea.describe()

count    2901.000000
mean     1503.191658
std       504.934510
min       334.000000
25%      1131.000000
50%      1445.000000
75%      1744.000000
max      5642.000000
Name: GrLivArea, dtype: float64

In [143]:
# Create a total "finished" basement square footage variable
df['BaseLivArea'] = df.TotalBsmtSF - df.BsmtUnfSF

C:\Users\Ahsan Salman Yousaf\AppData\Local\Temp\ipykernel_21368\736366612.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['BaseLivArea'] = df.TotalBsmtSF - df.BsmtUnfSF


In [144]:
# A simple linear regression of TSF on the SalePrice gives an R-squared of 50% by itself!
ols = LinearRegression()

X_GLA = df[['GrLivArea']]
y_SalePrice = df.SalePrice

ols.fit(X_GLA, y_SalePrice)
print("R-squared of Total Square Footage on Sale Price:", ols.score(X_GLA, y_SalePrice))

R-squared of Total Square Footage on Sale Price: 0.4989924240650856


In [145]:
df[['SalePrice', 'GrLivArea', 'BaseLivArea']].describe()

SalePrice    GrLivArea  BaseLivArea
count    2901.000000  2901.000000  2900.000000
mean   181827.768011  1503.191658   495.894483
std     79550.512426   504.934510   477.414744
min     12789.000000   334.000000     0.000000
25%    130000.000000  1131.000000     0.000000
50%    161500.000000  1445.000000   465.000000
75%    214000.000000  1744.000000   812.000000
max    755000.000000  5642.000000  5644.000000

### Searching for outliers

In [146]:
# There are 45 instances of 'SalePrice' being 3 std above the mean
df.SalePrice[(df.SalePrice > np.mean(df.SalePrice) + 3*np.std(df.SalePrice))].count()

45

In [147]:
# But there are zero cases of the opposite
df.SalePrice[(df.SalePrice < np.mean(df.SalePrice) - 3*np.std(df.SalePrice))].count()

0

In [148]:
# There are 25 instances of 'GrLivArea' being 3 std above the mean, but zero of the opposite
df.GrLivArea[(df.GrLivArea > np.mean(df.GrLivArea) + 3*np.std(df.GrLivArea))].count()

25

In [149]:
df.GrLivArea[(df.GrLivArea < np.mean(df.GrLivArea) - 3*np.std(df.GrLivArea))].count()

0

In [150]:
# There are 6 instances of 'BaseLivArea' being 3 std above the mean, but zero of the opposite
df.BaseLivArea[(df.BaseLivArea > np.mean(df.BaseLivArea) + 3*np.std(df.BaseLivArea))].count()

16

In [151]:
df.BaseLivArea[(df.BaseLivArea < np.mean(df.BaseLivArea) - 3*np.std(df.BaseLivArea))].count()

0

In [152]:
df.shape

(2901, 80)

In [153]:
df_exout = df[(df.SalePrice < np.mean(df.SalePrice) + 3*np.std(df.SalePrice))]

In [154]:
df_exout = df_exout[(df_exout.GrLivArea < np.mean(df.GrLivArea) + 3*np.std(df.GrLivArea))]

In [155]:
df_exout = df_exout[(df_exout.BaseLivArea < np.mean(df.BaseLivArea) + 3*np.std(df.BaseLivArea))]

In [156]:
# All in all, 66 outliers are deleted
df_exout.shape 

(2835, 80)

It's no brainer that the larger the house, the higher the same price. Real estate professionals tend to speak about a property's value in per unit terms, in this case it would be the price per square foot. So I construct a price per square foot measure using 'GrLivArea' as the denominator. In the US, the basement square footage is typically not cited in housing sames and surveyors tend not to count basement size to a house's official square footage.

In [157]:
df_exout['PriceSF'] = df_exout.SalePrice / df_exout.GrLivArea

### Searching for non-commercial transactions

In [158]:
print(df_exout.SalePrice.mean())
print(df_exout.PriceSF.mean())

175876.4934744268
120.88661412002561


In [159]:
# 113 of these "abnormal" sale transactions are listed as "abnormal" or to family members, 
# and their mean SalePrice is well below the overall mean SalePrice
print(df_exout.SalePrice[(df_exout.SaleCondition == 'Abnorml') | (df_exout.SaleCondition == 'Family')].count())
df_exout.SalePrice[(df_exout.SaleCondition == 'Abnorml') | (df_exout.SaleCondition == 'Family')].mean()

218


140131.2752293578

In [160]:
# Even the mean 'PriceSF' of these "abnormal" sales is well below the overall mean
df_exout.PriceSF[(df_exout.SaleCondition == 'Abnorml') | (df_exout.SaleCondition == 'Family')].mean()

103.08140598125344

I chose to exclude these 219 "abnormal" sale transactions from the analysis. The reason is that if these transactions are not fully commercial in nature, then their inclusion in the dataset will detract from the "economic" or commercial relationship between a property's characteristic and its price. In other words, these "abnormal" transactions will introduce bias into the model estimates.

In [161]:
df_exout = df_exout[(df_exout.SaleCondition != 'Abnorml')]
df_exout = df_exout[(df_exout.SaleCondition != 'Family')]
df_exout.shape

(2617, 81)

### Log transformation of the target variable

In [162]:
# Transform both sale price variables by taking the natural log to reduce the right-skew of the distributions
# The log transformation is also appropriate given that both variables have only non-zero positive values
df_exout['LnSalePrice'] = np.log(df_exout.SalePrice)
df_exout['LnPriceSF'] = np.log(df_exout.PriceSF)

Looking at the scatterplots of these key variables I have been working on, there are several notable observations. Although 'PriceSF' doesn't seem to a have a linear relationship with 'GrLivArea', it does appear to have a positive correlation with 'BaseLivArea'. This suggests that the size of the finished basement helps to raise 'PriceSF', which makes sense as it adds value to the property separate from 'GrLivArea'. Another interesting observation is that 'Age' seems to have more of a linear relationship with 'PriceSF' than with 'SalePrice'.

### Creating a "location" variable

In [163]:
# Examining the 'PriceSF' for the various neighbourhoods, it is obvious that the price premium differs significantly
# across them
neigh_mean = df_exout['PriceSF'].groupby(df_exout['Neighborhood']).count().sort_values()
neigh_mean

Neighborhood
Landmrk      1
GrnHill      2
Greens       8
Blueste     10
Veenker     21
NPkVill     22
BrDale      26
Blmngtn     28
MeadowV     34
StoneBr     38
ClearCr     38
SWISU       40
IDOTRR      57
NoRidge     60
Timber      66
Crawfor     94
BrkSide     98
Mitchel    103
NWAmes     117
SawyerW    118
Sawyer     137
NridgHt    141
Gilbert    158
Edwards    168
Somerst    173
OldTown    206
CollgCr    257
NAmes      396
Name: PriceSF, dtype: int64

In [164]:
len(neigh_mean)

28

In [165]:
# Creating a new numeric ordinal variable for 'Functional'
def functional_numeric(x):
    if 'Typ' in x:
        return 8
    elif 'Min1' in x:
        return 7
    elif 'Min2' in x:
        return 6
    elif 'Mod' in x:
        return 5
    elif 'Maj1' in x:
        return 4
    elif 'Maj2' in x:
        return 3
    elif 'Sev' in x:
        return 2    
    else:
        return 1
    
df_exout['Functional_Num'] = df_exout.Functional.map(functional_numeric)

In [166]:
# Creating a new numeric ordinal variable for external conditon or 'ExterCond'
def extercond_numeric(x):
    if 'Ex' in x:
        return 5
    elif 'Gd' in x:
        return 4
    elif 'TA' in x:
        return 3
    elif 'Fa' in x:
        return 2
    else:
        return 1
    
df_exout['ExterCond_Num'] = df_exout.ExterCond.map(extercond_numeric)

In [167]:
# Creating a new numeric ordinal variable for external quality or 'ExterQual'
def exterqual_numeric(x):
    if 'Ex' in x:
        return 5
    elif 'Gd' in x:
        return 4
    elif 'TA' in x:
        return 3
    elif 'Fa' in x:
        return 2
    else:
        return 1
    
df_exout['ExterQual_Num'] = df_exout.ExterQual.map(exterqual_numeric)

In [168]:
print(df_exout['OverallQual'].mean())
print(df_exout['OverallCond'].mean())
print(df_exout['ExterQual_Num'].mean())
print(df_exout['ExterCond_Num'].mean())
print(df_exout['Functional_Num'].mean())

6.103553687428353
5.595338173481085
3.4016048910966754
3.088651127244937
7.858616736721436


Construct the location desirability proxy from the 5 individual mean-standardised variables listed above. The construction methodology implies a stronger weighting to the external quality/condition of the building.

In [169]:
df_exout['Location'] = ((df_exout['OverallQual']/df_exout['OverallQual'].mean()) 
                        + (df_exout['OverallCond']/df_exout['OverallCond'].mean())
                        + (df_exout['ExterQual_Num']/df_exout['ExterQual_Num'].mean())
                        + (df_exout['ExterCond_Num']/df_exout['ExterCond_Num'].mean()) 
                        + (df_exout['Functional_Num']/df_exout['Functional_Num'].mean()))

In [170]:
df_exout.Location.describe()

count    2617.000000
mean        5.000000
std         0.437242
min         2.087044
25%         4.747860
50%         5.041839
75%         5.269140
max         7.353587
Name: Location, dtype: float64

In [171]:
df_exout.Location.median()

5.041839076469422

In [172]:
# The average 'Location' scores of the respective neighbourhoods
df_exout['Location'].groupby([df_exout.Neighborhood]).mean().sort_values()

Neighborhood
MeadowV    4.616332
Edwards    4.687926
Sawyer     4.743020
Landmrk    4.747860
SWISU      4.754528
BrDale     4.778427
IDOTRR     4.787411
NAmes      4.824162
BrkSide    4.836702
Mitchel    4.846688
NPkVill    4.911514
OldTown    4.947847
ClearCr    4.954413
Gilbert    5.013664
SawyerW    5.020900
NWAmes     5.038560
Crawfor    5.132186
CollgCr    5.138734
Blueste    5.145065
GrnHill    5.205678
Blmngtn    5.229084
Timber     5.232063
Somerst    5.267111
NoRidge    5.380242
StoneBr    5.404945
Greens     5.414197
Veenker    5.463750
NridgHt    5.469528
Name: Location, dtype: float64

In [173]:
df_exout['SalePrice'].groupby([df_exout.Neighborhood]).count().sort_values()

Neighborhood
Landmrk      1
GrnHill      2
Greens       8
Blueste     10
Veenker     21
NPkVill     22
BrDale      26
Blmngtn     28
MeadowV     34
StoneBr     38
ClearCr     38
SWISU       40
IDOTRR      57
NoRidge     60
Timber      66
Crawfor     94
BrkSide     98
Mitchel    103
NWAmes     117
SawyerW    118
Sawyer     137
NridgHt    141
Gilbert    158
Edwards    168
Somerst    173
OldTown    206
CollgCr    257
NAmes      396
Name: SalePrice, dtype: int64

From the analysis above, there is a possible way to construct an ordinal measurement of locational desirability by using the percentiles of the distribution of the new 'Location' variable. I decided that with no additional information or insight, I should just allocate 4 ordinal values for 'Location': 1 (low), 2 (mid-low), 3 (mid-high) and 4 (high).

First, I sorted the 'Neighborhood' variable by the mean 'Location' values, then slice it in half at approximately the mean (5.00), so that there are 14 'Neighborhood' instances below and 14 above this arbitrary division. This resulted in all neighbourhoods above "SawyerW" (as listed above), which had a mean 'Location' reading of 5.02, to fall into the 1 and 2 readings of 'Location'. I then arbitrary assigned 'Location' = 1 for the 7 lowest neighbourhoods by their mean 'Location' values, and 'Location' = 2 for the remaining 7.

As for the fourteen neighbourhoods in the upper half of the 'Location' values, I divided them exactly in half again, with the lower 7 in terms of their mean 'Location' values assigned 'Location' = 3, and the remainder assigned 'Location' = 4. 

In this way, I have created an ordinal 'Location' desirability score of 1 to 4. The number of observations in the data also do not appear to be too unbalanced across these four ranks, though 61% of the observations are in the first two 'Location' categories. That said, there was a right skew to the 'SalePrice' as discussed above, so there will likely be a smaller number of expensive neighborhoods.

In [174]:
def add_location(x):
    if 'MeadowV' in x or 'Edwards' in x or 'Sawyer' in x or 'Landmrk' in x or 'SWISU' in x or 'BrDale' in x or 'IDOTRR' in x:
        return 1
    elif 'NAmes' in x or 'Mitchel' in x or 'BrkSide' in x or 'NPkVill' in x or 'OldTown' in x or 'ClearCr' in x or 'Gilbert' in x:
        return 2
    elif 'SawyerW' in x or 'NWAmes' in x or 'Crawfor' in x or 'CollgCr' in x or 'Blueste' in x or 'GrnHill' in x or 'Blmngtn' in x:
        return 3
    else:
        return 4

In [175]:
df_exout['Location'] = df_exout.Neighborhood.map(add_location)

In [176]:
df_exout['Location'].value_counts()

Location
2    1021
1     581
3     508
4     507
Name: count, dtype: int64

In [177]:
# Positive correlation between 'SalePrice' and 'PriceSF' with 'Location'
print(df_exout['SalePrice'].groupby(df_exout.Location).mean())
print(df_exout['PriceSF'].groupby(df_exout.Location).mean())

Location
1    138194.593804
2    151184.290891
3    198714.606299
4    261270.218935
Name: SalePrice, dtype: float64
Location
1    108.533247
2    113.431800
3    129.737730
4    148.842905
Name: PriceSF, dtype: float64


In [178]:
# Total mean house size also rises with 'Location'
print(df_exout['GrLivArea'].groupby(df_exout.Location).mean())
print(df_exout['GrLivArea'].groupby(df_exout.Location).median())

Location
1    1315.370052
2    1376.113614
3    1567.360236
4    1775.142012
Name: GrLivArea, dtype: float64
Location
1    1214.0
2    1344.0
3    1565.5
4    1692.0
Name: GrLivArea, dtype: float64


So it does appear that the 'Location' score does tend to be positively correlated with house prices, and somewhat with property sizes too.

# Other feature engineering on sale year, seasonality, zoning, and proximinity to railways & artery roads

In [179]:
df_exout['YrSold'] = df_exout['YrSold'].astype(np.int64)

In [180]:
# Looking at the 'SalePrice' and 'PriceSF' grouped by 'YrSold', there is an obvious impact of the financial crisis 
# that hit in 2007-2008. The mean 'SoldPrice' and 'PriceSF' in 2009 and 2010 are still lower than those in 2007.
print(df_exout.SalePrice.groupby(df_exout.YrSold).mean())
print(df_exout.PriceSF.groupby(df_exout.YrSold).mean())

YrSold
2006    180405.762963
2007    179753.426540
2008    179122.688969
2009    178999.132302
2010    173546.498382
Name: SalePrice, dtype: float64
YrSold
2006    121.286039
2007    123.869837
2008    122.638253
2009    121.761175
2010    121.856891
Name: PriceSF, dtype: float64


In [181]:
# House prices were in an uptrend in 2006-2007, but fell in 2008 and have been moribund since. 
df_exout.PriceSF.groupby(df_exout.YrSold).mean().pct_change()

YrSold
2006         NaN
2007    0.021303
2008   -0.009943
2009   -0.007152
2010    0.000786
Name: PriceSF, dtype: float64

In [182]:
# Thus, it might be useful to incorporate year dummies in the regression model
df_exout = pd.get_dummies(df_exout, columns=['YrSold'])

In [183]:
# Clear price differentials according to the discrete housing zones
df_exout.SalePrice.groupby(df_exout.MSZoning).mean()

MSZoning
FV    220161.447761
RH    130250.000000
RL    186635.804284
RM    128617.762836
Name: SalePrice, dtype: float64

In [184]:
# Creating a new ordinal variable of zoning, corresponding to the mean values from 'MSZoning'
# The FV observations in 'MSZoning' are a retirement community development, named "Floating Village", which appears 
# wealthy retireesto cater to

def add_zoning(x):
    if 'RM' in x:
        return 1
    elif 'RH' in x:
        return 2
    elif 'RL' in x:
        return 3
    else:
        return 4

In [185]:
df_exout['Zoning'] = df_exout.MSZoning.map(add_zoning)

In [186]:
# No strong linear relationship between 'GrLivArea' or 'BaseLivArea' and 'Zoning'
print(df_exout.GrLivArea.groupby(df_exout.Zoning).mean())
df_exout.BaseLivArea.groupby(df_exout.Zoning).mean()

Zoning
1    1312.369193
2    1398.450000
3    1503.565725
4    1585.119403
Name: GrLivArea, dtype: float64


Zoning
1    278.310513
2    420.800000
3    524.361733
4    385.507463
Name: BaseLivArea, dtype: float64

In [187]:
# There appears to be a sizeable jump in prices in the low density and FV categories
print(df_exout.SalePrice.groupby(df_exout.Zoning).mean())
df_exout.PriceSF.groupby(df_exout.Zoning).mean()

Zoning
1    128617.762836
2    130250.000000
3    186635.804284
4    220161.447761
Name: SalePrice, dtype: float64


Zoning
1    102.015245
2     98.920928
3    125.501798
4    139.988488
Name: PriceSF, dtype: float64

In [188]:
# There is a relationship between 'Location' and 'Zoning' with Locations 3 and 4 
# primarily in zones 3 and 4 (low density and the retirement village)
df_exout.Location.groupby(df_exout.Zoning).value_counts()

Zoning  Location
1       2           240
        1           131
        3            37
        4             1
2       1            12
        2             6
        3             2
3       2           775
        3           469
        1           438
        4           372
4       4           134
Name: count, dtype: int64

In [189]:
# Analysis above suggests that it might be more useful to model 'Zoning' using dummies instead of an ordinal variable
df_exout['Zone_ordinal'] = df_exout['Zoning']
df_exout = pd.get_dummies(df_exout, columns=['Zoning'])

In [190]:
# The overwhelming majority of observations fall into the double-'Norm' category in terms of
# Condition1 and Condition2, which is a measure of environmental condition
df_exout.Condition1.groupby(df_exout.Condition2).value_counts()

Condition2  Condition1
Artery      RRNn             1
            Artery           1
Feedr       RRAn             8
            Feedr            3
            RRNn             1
Norm        Norm          2263
            Feedr          136
            Artery          75
            RRAn            37
            PosN            32
            RRAe            25
            PosA            17
            RRNn             6
            RRNe             4
PosA        Artery           1
            PosA             1
PosN        PosN             2
RRAe        Feedr            1
RRAn        Feedr            1
RRNn        Feedr            2
Name: count, dtype: int64

In [191]:
# Previous studies have shown a negative relationship between housing prices and being adjacent 
# or near to a major roadway or a railway line.
# Creating variables to capture this negative enviromental impact
def add_roadrail1(x):
    if 'Artery' in x:
        return 1
    elif 'RRAn' in x:
        return 1
    elif 'RRNn' in x:
        return 1
    elif 'RRAe' in x:
        return 1
    elif 'RRNe' in x:
        return 1
    else:
        return 0

df_exout['RoadRail1'] = df_exout.Condition1.map(add_roadrail1)

In [192]:
def add_roadrail2(x):
    if 'Artery' in x:
        return 1
    elif 'RRAn' in x:
        return 1
    elif 'RRNn' in x:
        return 1
    elif 'RRAe' in x:
        return 1
    elif 'RRNe' in x:
        return 1
    else:
        return 0

df_exout['RoadRail2'] = df_exout.Condition2.map(add_roadrail2)

In [193]:
df_exout.RoadRail1.groupby(df_exout.RoadRail2).value_counts()

RoadRail2  RoadRail1
0          0            2454
           1             157
1          0               4
           1               2
Name: count, dtype: int64

In [194]:
# Combining the 2 proximity to artery roads and railroad dummy variables into one
df_exout['RoadRail'] = df_exout.apply(lambda x: 1 if (x['RoadRail1'] == 1 | x['RoadRail2'] == 1) 
                                      else 0, axis=1)
df_exout.drop(['RoadRail1', 'RoadRail2'], axis=1, inplace=True)

In [195]:
# The RoadRail variable seems to have an impact on mean 'SalePrice' and 'PriceSF' in every 'Location'
print(df_exout.SalePrice.groupby([df_exout.RoadRail, df_exout.Location]).mean())
df_exout.PriceSF.groupby([df_exout.RoadRail, df_exout.Location]).mean()

RoadRail  Location
0         1           139132.013060
          2           152111.176915
          3           199110.281124
          4           261031.189135
1         1           127028.888889
          2           142043.617021
          3           179010.000000
          4           273150.000000
Name: SalePrice, dtype: float64


RoadRail  Location
0         1           108.666221
          2           114.734104
          3           130.069818
          4           148.727217
1         1           106.949380
          2           100.588870
          3           113.199754
          4           154.592604
Name: PriceSF, dtype: float64

### More (minor) feature engineering

In [196]:
# Transforming the 'CentralAir' discrete variable to numeric
df_exout['CentralAirNum'] = df_exout.apply(lambda x: 1 if (x['CentralAir'] == 'Y') 
                                                           else 0, axis=1)
#df_exout[['CentralAir', 'CentralAirNum']]
df_exout['CentralAirNum'].value_counts()

CentralAirNum
1    2465
0     152
Name: count, dtype: int64

In [197]:
# Creating variables on positive neighbourhood amenities
def add_amenities1(x):
    if 'PosN' in x:
        return 1
    elif 'PosA' in x:
        return 1
    else:
        return 0

df_exout['Amenities1'] = df_exout.Condition1.map(add_amenities1)

In [198]:
def add_amenities2(x):
    if 'PosN' in x:
        return 1
    elif 'PosA' in x:
        return 1
    else:
        return 0

df_exout['Amenities2'] = df_exout.Condition2.map(add_amenities2)

In [199]:
# Combining the amenities dummy variables into one
df_exout['Amenities'] = df_exout.apply(lambda x: 1 if (x['Amenities1'] == 1 | x['Amenities2'] == 1) 
                                      else 0, axis=1)
df_exout.drop(['Amenities1', 'Amenities2'], axis=1, inplace=True)

In [200]:
# 'Amenities'=1 does seem to raise the mean 'SalePrice' but has mixed impact on 'PriceSF'
print(df_exout.SalePrice.groupby([df_exout.Amenities, df_exout.Location]).mean())
df_exout.PriceSF.groupby([df_exout.Amenities, df_exout.Location]).mean()

Amenities  Location
0          1           137982.801038
           2           150879.137041
           3           197069.480167
           4           260569.862275
1          1           179000.000000
           2           173133.571429
           3           225887.551724
           4           319750.000000
Name: SalePrice, dtype: float64


Amenities  Location
0          1           108.385444
           2           113.546139
           3           130.210397
           4           148.480976
1          1           137.009921
           2           105.207578
           3           121.930565
           4           179.063978
Name: PriceSF, dtype: float64

In [201]:
# At this point, all the observations in the sample have full public utility service
df_exout[(df_exout.Utilities != 'AllPub')].count()

PID              1
MSSubClass       1
MSZoning         1
LotFrontage      1
LotArea          1
                ..
Zoning_3         1
Zoning_4         1
RoadRail         1
CentralAirNum    1
Amenities        1
Length: 99, dtype: int64

In [202]:
# Adding up the above ground bathrooms and assigning to a new variable
df_exout['Bathrooms'] = df_exout.FullBath + (0.5*df_exout.HalfBath)
df_exout.Bathrooms.describe()

count    2617.000000
mean        1.757738
std         0.628060
min         0.000000
25%         1.000000
50%         2.000000
75%         2.500000
max         5.000000
Name: Bathrooms, dtype: float64

In [203]:
# Positive relationship between the overall quality of the housing structure and the 'SalePrice' and 'PriceSF'
print(df_exout.SalePrice.groupby([df_exout.OverallQual]).mean())
df_exout.PriceSF.groupby([df_exout.OverallQual]).mean()

OverallQual
1.0      50150.000000
2.0      58667.111111
3.0      87770.689655
4.0     109863.445652
5.0     135816.621769
6.0     162948.944776
7.0     205672.802102
8.0     266258.361111
9.0     341216.761905
10.0    366957.333333
Name: SalePrice, dtype: float64


OverallQual
1.0      92.571273
2.0      92.689487
3.0      90.229760
4.0     100.208193
5.0     114.006923
6.0     116.030656
7.0     126.365056
8.0     147.828978
9.0     178.511137
10.0    176.209673
Name: PriceSF, dtype: float64

In [204]:
# Interestingly 'OverallQual' is not well correlated with 'OverallCond'
df_exout.OverallQual.corr(df_exout.OverallCond)

-0.13594329570462232

In [205]:
# But it is moderately correlated with the 'Location' variable as one would expect
df_exout.OverallQual.corr(df_exout.Location)

0.6253793053742019

In [206]:
# Overall Condition of the house appears to have mixed impact on prices
print(df_exout.SalePrice.groupby([df_exout.OverallCond]).mean())
df_exout.PriceSF.groupby([df_exout.OverallCond]).mean()

OverallCond
1.0     58000.000000
2.0    142261.714286
3.0     95744.138889
4.0    119924.219512
5.0    201306.443396
6.0    151080.516667
7.0    154871.446023
8.0    153281.834532
9.0    188300.000000
Name: SalePrice, dtype: float64


OverallCond
1.0     59.061222
2.0     88.624891
3.0     80.767567
4.0     89.005222
5.0    129.065319
6.0    115.397512
7.0    118.068083
8.0    118.683092
9.0    123.331215
Name: PriceSF, dtype: float64

### Dummifying other potential important features

In [207]:
df_exout = pd.get_dummies(df_exout, columns=['Street'])

In [208]:
df_exout.drop(['Street_Grvl'], axis=1, inplace=True)
df_exout.rename(columns={'Street_Pave': 'Street_dum'}, inplace=True)

In [209]:
df_exout = pd.get_dummies(df_exout, columns=['KitchenQual'])

In [210]:
# Creating a dummy variable for flat roofs
df_exout['FlatRoof_dum'] = df_exout.RoofStyle.apply(lambda x: 1 if x=='Flat' else 0)

In [211]:
# Creating a dummy variable for garage
df_exout['Garage_dum'] = df_exout.GarageQual.apply(lambda x: 0 if pd.isnull(x)==True else 1)

In [212]:
# Creating a dummy variable for flat property contour
df_exout['FlatContour_dum'] = df_exout.LandContour.apply(lambda x: 1 if x=='Lvl' else 0)

In [213]:
# Creating a dummy variable for houses higher than one storey
df_exout['TwoStory_dum'] = df_exout['2ndFlrSF'].apply(lambda x: 1 if x>0 else 0)

In [214]:
df_exout.shape

(2617, 108)

# Finalising the dataset

In [215]:
pd.options.display.max_columns = None
df_exout.head()

PID  MSSubClass MSZoning  LotFrontage  LotArea LotShape LandContour  \
0  526301100        20.0       RL        141.0  31770.0      IR1         Lvl   
1  526350040        20.0       RH         80.0  11622.0      Reg         Lvl   
2  526351010        20.0       RL         81.0  14267.0      IR1         Lvl   
3  526353030        20.0       RL         93.0  11160.0      Reg         Lvl   
4  527105010        60.0       RL         74.0  13830.0      IR1         Lvl   

  Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam   
1    AllPub    Inside       Gtl        NAmes      Feedr       Norm     1Fam   
2    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam   
3    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam   
4    AllPub    Inside       Gtl      Gilbert       Norm       Norm     1Fam   

  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemod/Add RoofStyle  \
0     1Story          6.0          5.0     1960.0         1960.0       Hip   
1     1Story          5.0          6.0     1961.0         1961.0     Gable   
2     1Story          6.0          6.0     1958.0         1958.0       Hip   
3     1Story          7.0          5.0     1968.0         1968.0       Hip   
4     2Story          5.0          5.0     1997.0         1998.0     Gable   

  RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond  \
0  CompShg     BrkFace     Plywood      Stone       112.0        TA        TA   
1  CompShg     VinylSd     VinylSd        NaN         0.0        TA        TA   
2  CompShg     Wd Sdng     Wd Sdng    BrkFace       108.0        TA        TA   
3  CompShg     BrkFace     BrkFace        NaN         0.0        Gd        TA   
4  CompShg     VinylSd     VinylSd        NaN         0.0        TA        TA   

  Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1  \
0     CBlock       TA       Gd           Gd          BLQ       639.0   
1     CBlock       TA       TA           No          Rec       468.0   
2     CBlock       TA       TA           No          ALQ       923.0   
3     CBlock       TA       TA           No          ALQ      1065.0   
4      PConc       Gd       TA           No          GLQ       791.0   

  BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC  \
0          Unf         0.0      441.0       1080.0    GasA        Fa   
1          LwQ       144.0      270.0        882.0    GasA        TA   
2          Unf         0.0      406.0       1329.0    GasA        TA   
3          Unf         0.0     1045.0       2110.0    GasA        Ex   
4          Unf         0.0      137.0        928.0    GasA        Gd   

  CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  \
0          Y      SBrkr    1656.0       0.0           0.0     1656.0   
1          Y      SBrkr     896.0       0.0           0.0      896.0   
2          Y      SBrkr    1329.0       0.0           0.0     1329.0   
3          Y      SBrkr    2110.0       0.0           0.0     2110.0   
4          Y      SBrkr     928.0     701.0           0.0     1629.0   

   BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0           1.0           0.0       1.0       0.0           3.0           1.0   
1           0.0           0.0       1.0       0.0           2.0           1.0   
2           0.0           0.0       1.0       1.0           3.0           1.0   
3           1.0           0.0       2.0       1.0           3.0           1.0   
4           0.0           0.0       2.0       1.0           3.0           1.0   

   TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt  \
0           7.0        Typ         2.0          Gd     Attchd       1960.0   
1           5.0        Typ         0.0         NaN     Attchd       1961.0   
2           6.0        Typ         0.0         NaN     Attchd       1958.0   

In [216]:
df_exout = df_exout[['SalePrice', 'LnSalePrice', 'Age', 'GrLivArea', 'BaseLivArea', 'Location', 'Amenities', 
                     'RoadRail', 'BedroomAbvGr', 'Bathrooms', 'OverallCond', 'OverallQual', 'LotFrontage', 
                     'LotArea', 'TwoStory_dum', 'FlatContour_dum', 'FlatRoof_dum', 'GarageArea', 'Garage_dum', 
                     'CentralAirNum', 'LowQualFinSF', 'Fireplaces', 'KitchenQual_Ex', 'Zoning_2', 'Zoning_3', 
                     'Zoning_4', 'YrSold_2007', 'YrSold_2008', 'YrSold_2009', 'YrSold_2010']]

In [217]:
# Saving engineered dataset
df_exout.to_csv(r'data_engineered.csv', index=False)